data/scripts.js

```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		if (!targets.length) {
+			this.attrLastMove('[notarget]');
+			this.add(this.gen >= 5 ? '-fail' : '-notarget', pokemon);
+			return false;
+		}
+		if (targets.length > 1) move.spreadHit = true;
+
+		/** @type {((targets: Pokemon[], pokemon: Pokemon, move: ActiveMove) => (number | boolean | "" | undefined)[] | undefined)[]} */
+		let moveSteps = [
+			// 0. check for semi invulnerability
+			this.tryImmunityEvent,
+
+			// 1. run the 'TryHit' event (Protect, Magic Bounce, Volt Absorb, etc.) (this is step 2 in gens 5 & 6, and step 5 in gen 4)
+			this.tryHitEvent,
+
+			// 2. check for type immunity (this is step 1 in gens 4-6)
+			this.typeImmunity,
+
+			// 3. check for powder immunity (gen 6+ only)
+			this.powderImmunity,
+
+			// 4. check for prankster immunity (gen 6+ only)
+			this.pranksterImmunity,
+
+			// 5. check accuracy
+			this.accuracy,
+
+			// 6. break protection effects
+			this.breakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.stealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.moveHitLoop,
+		];
+		if (this.gen <= 6) {
+			// Swap step 1 with step 2
+			[moveSteps[1], moveSteps[2]] = [moveSteps[2], moveSteps[1]];
+		}
+		if (this.gen === 4) {
+			// Swap step 5 with new step 2 (old step 1)
+			[moveSteps[2], moveSteps[5]] = [moveSteps[5], moveSteps[2]];
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
 		move.zBrokeProtect = false;
 
-		let hitResult = this.singleEvent('PrepareHit', move, {}, target, pokemon, move);
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
 		if (!hitResult) {
 			if (hitResult === false) {
 				this.add('-fail', pokemon);
 				this.attrLastMove('[still]');
 			}
 			return false;
 		}
-		this.runEvent('PrepareHit', pokemon, target, move);
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
 
-		if (!this.singleEvent('Try', move, null, pokemon, target, move)) {
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
 			return false;
 		}
 
-		if (move.target === 'all' || move.target === 'foeSide' || move.target === 'allySide' || move.target === 'allyTeam') {
-			if (move.target === 'all') {
-				hitResult = this.runEvent('TryHitField', target, pokemon, move);
-			} else {
-				hitResult = this.runEvent('TryHitSide', target, pokemon, move);
-			}
-			if (!hitResult) {
-				if (hitResult === false) {
-					this.add('-fail', pokemon);
-					this.attrLastMove('[still]');
-				}
-				return false;
+		let atLeastOneFailure;
+		for (const step of moveSteps) {
+			/** @type {(number | boolean | "" | undefined)[] | undefined} */
+			let hitResults = step.call(this, targets, pokemon, move);
+			if (!hitResults) continue;
+			// @ts-ignore
+			targets = targets.filter((val, i) => hitResults[i] || hitResults[i] === 0);
+			atLeastOneFailure = atLeastOneFailure || hitResults.some(val => val === false);
+			if (!targets.length) {
+				// console.log(step.name);
```

Comment:
remove commented out log and just make this one line?

Reply:
It's a useful breakpoint point; I'm fine with leaving it in.